File: examples/browser/serve.js
#!/usr/bin/env node

/**
 * Simple HTTP server for testing the Sockudo WebSocket client
 * Run with: node serve.js
 */

const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = 8080;
const HOST = 'localhost';

// MIME types
const mimeTypes = {
    '.html': 'text/html',
    '.js': 'text/javascript',
    '.wasm': 'application/wasm',
    '.css': 'text/css',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
    '.ico': 'image/x-icon'
};

const server = http.createServer((req, res) => {
    console.log(`${new Date().toISOString()} - ${req.method} ${req.url}`);

    // Parse URL
    let filePath = req.url === '/' ? '/test.html' : req.url;

    // Remove query string
    filePath = filePath.split('?')[0];

    // Security: prevent directory traversal
    if (filePath.includes('..')) {
        res.writeHead(403, { 'Content-Type': 'text/plain' });
        res.end('403 Forbidden');
        return;
    }

    // Map URLs to file paths
    let fullPath;
    if (filePath.startsWith('/pkg/')) {
        // Serve WASM package files from ../../pkg/
        fullPath = path.join(__dirname, '../../pkg', filePath.substring(5));
    } else {
        // Serve HTML/CSS/JS from current directory
        fullPath = path.join(__dirname, filePath);
    }

    // Get file extension
    const ext = path.extname(fullPath).toLowerCase();
    const contentType = mimeTypes[ext] || 'application/octet-stream';

    // Read and serve file
    fs.readFile(fullPath, (err, content) => {
        if (err) {
            if (err.code === 'ENOENT') {
                res.writeHead(404, { 'Content-Type': 'text/html' });
                res.end(`
                    <!DOCTYPE html>
                    <html>
                    <head><title>404 Not Found</title></head>
                    <body>
                        <h1>404 - File Not Found</h1>
                        <p>The requested file <code>${filePath}</code> was not found.</p>
                        <p>Looking for: <code>${fullPath}</code></p>
                        <a href="/">Go back to home</a>
                    </body>
                    </html>
                `);
            } else {
                res.writeHead(500, { 'Content-Type': 'text/plain' });
                res.end(`500 Internal Server Error: ${err.code}`);
            }
        } else {
            // Add CORS headers for WASM
            res.writeHead(200, {
                'Content-Type': contentType,
                'Cross-Origin-Opener-Policy': 'same-origin',
                'Cross-Origin-Embedder-Policy': 'require-corp',
                'Access-Control-Allow-Origin': '*'
            });
            res.end(content);
        }
    });
});

server.listen(PORT, HOST, () => {
    console.log('='.repeat(60));
    console.log('🚀 Sockudo Test Server Started!');
    console.log('='.repeat(60));
    console.log(`📍 Server running at: http://${HOST}:${PORT}/`);
    console.log(`📄 Test page: http://${HOST}:${PORT}/test.html`);
    console.log('');
    console.log('Press Ctrl+C to stop the server');
    console.log('='.repeat(60));
});

// Graceful shutdown
process.on('SIGINT', () => {
    console.log('\n\n👋 Shutting down server...');
    server.close(() => {
        console.log('✅ Server closed');
        process.exit(0);
    });
});

process.on('SIGTERM', () => {
    console.log('\n\n👋 Shutting down server...');
    server.close(() => {
        console.log('✅ Server closed');
        process.exit(0);
    });
});

File: examples/browser/serve.py
#!/usr/bin/env python3

"""
Simple HTTP server for testing the Sockudo WebSocket client
Run with: python serve.py
"""

import http.server
import socketserver
import os
import sys
from pathlib import Path

PORT = 8080
HOST = "localhost"

class CORSRequestHandler(http.server.SimpleHTTPRequestHandler):
    """HTTP request handler with CORS support for WASM"""

    def __init__(self, *args, **kwargs):
        # Set the directory to serve files from
        super().__init__(*args, directory=str(Path(__file__).parent), **kwargs)

    def end_headers(self):
        # Add CORS headers
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', '*')
        self.send_header('Cross-Origin-Opener-Policy', 'same-origin')
        self.send_header('Cross-Origin-Embedder-Policy', 'require-corp')
        super().end_headers()

    def guess_type(self, path):
        """Add WASM MIME type support"""
        if path.endswith('.wasm'):
            return 'application/wasm'
        elif path.endswith('.js'):
            return 'text/javascript'
        return super().guess_type(path)

    def translate_path(self, path):
        """Translate URL path to filesystem path, handling /pkg/ specially"""
        # Remove query string
        path = path.split('?', 1)[0]
        path = path.split('#', 1)[0]

        # Default to test.html for root
        if path == '/':
            path = '/test.html'

        # Handle /pkg/ URLs - serve from ../../pkg/
        if path.startswith('/pkg/'):
            pkg_path = Path(__file__).parent.parent.parent / 'pkg' / path[5:]
            return str(pkg_path)

        # Serve from current directory
        return super().translate_path(path)

    def log_message(self, format, *args):
        """Custom log format"""
        sys.stdout.write(f"[{self.log_date_time_string()}] {format % args}\n")


def run_server():
    """Start the HTTP server"""
    handler = CORSRequestHandler

    with socketserver.TCPServer((HOST, PORT), handler) as httpd:
        print("=" * 60)
        print("🚀 Sockudo Test Server Started!")
        print("=" * 60)
        print(f"📍 Server running at: http://{HOST}:{PORT}/")
        print(f"📄 Test page: http://{HOST}:{PORT}/test.html")
        print()
        print("Press Ctrl+C to stop the server")
        print("=" * 60)

        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\n\n👋 Shutting down server...")
            httpd.shutdown()
            print("✅ Server closed")


if __name__ == "__main__":
    # Check if pkg directory exists
    pkg_dir = Path(__file__).parent.parent.parent / 'pkg'
    if not pkg_dir.exists():
        print("⚠️  WARNING: pkg directory not found!")
        print("Please run the WASM build first:")
        print("  ./build-wasm.sh")
        print()
        response = input("Continue anyway? (y/N): ")
        if response.lower() != 'y':
            sys.exit(1)

    run_server()
